src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -348,6 +364,48 @@ private void EventLoop()
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ScheduleToProcessEvents()
+        {
+            // Schedule a thread pool work item to process events. Only one work item is scheduled at any given time to avoid
+            // over-parallelization. When the work item begins running, this field is reset to 0, allowing for another work item
+            // to be scheduled for parallelizing processing of events.
+            if (Interlocked.CompareExchange(ref _eventQueueProcessingRequested, 1, 0) == 0)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+            }
+        }
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            // Indicate that a work item is no longer scheduled to process events. The change needs to be visible to enqueuer
+            // threads (only for EventLoop() currently) before an event is attempted to be dequeued. In particular, if an
+            // enqueuer queues an event and does not schedule a work item because it is already scheduled, and this thread is
+            // the last thread processing events, it must see the event queued by the enqueuer.
+            Interlocked.Exchange(ref _eventQueueProcessingRequested, 0);
+
+            ConcurrentDictionary<IntPtr, SocketAsyncContext> handleToContextMap = _handleToContextMap;
+            ConcurrentQueue<Event> eventQueue = _eventQueue;
+            while (eventQueue.TryDequeue(out Event ev))
+            {
+                handleToContextMap.TryGetValue(ev.Handle, out SocketAsyncContext? context);
+                if (context == null)
+                {
+                    continue;
+                }
+
+                // An event was successfully dequeued, and as there may be more events to process, speculatively schedule a work
+                // item to parallelize processing of events. Since this is only for additional parallelization, doing so
+                // speculatively is ok.
+                if (_eventQueueProcessingRequested == 0)
```

Comment:
Any benefit to checking IsEmpty on the queue and only scheduling if it's not empty? Or we expect the common case will be that more work is available, and if it isn't the overhead of the extra work item won't matter? IsEmpty isn't free, so there's a trade-off; using it might be the wrong one.

Reply:
I expect it wouldn't be worth the overhead. In heavily loaded cases the queue usually wouldn't be empty. In lightly loaded cases there probably wouldn't be a significant perf difference, though with small additional CPU time from queuing and running the extra work item.

Reply:
FWIW `IsEmpty` is more expensive on `ARM` compared to `x64`.